cmd/vtysh.go

```diff
@@ -0,0 +1,155 @@
+package cmd
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os/exec"
+	"strings"
+
+	"k8s.io/klog/v2"
+)
+
+type (
+	BGPSummary struct {
+		Ipv4Unicast Ipv4Unicast `json:"ipv4Unicast"`
+	}
+	Ipv4Unicast struct {
+		RouterID    string          `json:"routerId"`
+		Peers       map[string]Peer `json:"peers"`
+		FailedPeers int             `json:"failedPeers"`
+		TotalPeers  int             `json:"totalPeers"`
+	}
+	Peer struct {
+		Hostname                   string `json:"hostname"`
+		RemoteAs                   int64  `json:"remoteAs"`
+		LocalAs                    int64  `json:"localAs"`
+		Version                    int    `json:"version"`
+		MsgRcvd                    int    `json:"msgRcvd"`
+		MsgSent                    int    `json:"msgSent"`
+		TableVersion               int    `json:"tableVersion"`
+		Outq                       int    `json:"outq"`
+		Inq                        int    `json:"inq"`
+		PeerUptime                 string `json:"peerUptime"`
+		PeerUptimeMsec             int    `json:"peerUptimeMsec"`
+		PeerUptimeEstablishedEpoch int    `json:"peerUptimeEstablishedEpoch"`
+		PfxRcd                     int    `json:"pfxRcd"`
+		PfxSnt                     int    `json:"pfxSnt"`
+		State                      string `json:"state"`
+		PeerState                  string `json:"peerState"`
+		ConnectionsEstablished     int    `json:"connectionsEstablished"`
+		ConnectionsDropped         int    `json:"connectionsDropped"`
+		IDType                     string `json:"idType"`
+	}
+)
+
+func repairFailedBGPSession() error {
+	failedInterfaces, err := checkBGPSessions("lan")
+	if err != nil {
+		return err
+	}
+
+	for _, name := range failedInterfaces {
+		klog.Infof("bgp session on interface %q is broken, try to repair", name)
+		err := ethtool(name)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func checkBGPSessions(interfaceFilter string) ([]string, error) {
+	socketPath, err := lookupSocketPath("bgpd")
+	if err != nil {
+		return nil, err
+	}
+	output, err := runCmd(socketPath, "show bgp ipv4 summary json")
+	if err != nil {
+		return nil, err
+	}
+
+	var bgpSummary BGPSummary
+	err = json.Unmarshal(output, &bgpSummary)
+	if err != nil {
+		return nil, err
+	}
+
+	if bgpSummary.Ipv4Unicast.FailedPeers == 0 {
+		return nil, nil
+	}
+
+	failedInterfaces := []string{}
+	for name, peer := range bgpSummary.Ipv4Unicast.Peers {
+		if !strings.HasPrefix(name, interfaceFilter) {
+			continue
+		}
+		// see https://github.com/FRRouting/frr/blob/stable/8.2/bgpd/bgp_debug.c#L102
+		if peer.State == "Idle" {
```

Comment:
I have seen (but unfortunately not documented :sob:) the bgp session being stuck in newly provisioned machines that were in state "Wating" when the switch reinstall happened. With these machines that state should be "Connect", not "Idle" - maybe you can add that state too?

Reply:
done